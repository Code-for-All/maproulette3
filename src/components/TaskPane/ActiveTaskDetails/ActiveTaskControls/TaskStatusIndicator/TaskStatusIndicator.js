import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { FormattedMessage } from 'react-intl'
import { messagesByStatus }
       from '../../../../../services/Task/TaskStatus/TaskStatus'
import WithDeactivateOnOutsideClick
       from '../../../../HOCs/WithDeactivateOnOutsideClick/WithDeactivateOnOutsideClick'
import Popout from '../../../../Bulma/Popout'
import SvgSymbol from '../../../../SvgSymbol/SvgSymbol'
import messages from './Messages'
import './TaskStatusIndicator.css'

const DeactivatablePopout = WithDeactivateOnOutsideClick(Popout)

/**
 * TaskStatusIndicator displays the current status of the given task.
 * If isMinimized is set to true, then it makes use of a popout component
 * with an icon control.
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
export default class TaskStatusIndicator extends Component {
  render() {
    if (this.props.isMinimized) {
      const popoutButton = (
        <button className="button icon-only task-status">
          <span className="control-icon"
                title={this.props.intl.formatMessage(messages.statusTooltip)}>
            <SvgSymbol viewBox='0 0 20 20' sym="flag-icon" />
          </span>
        </button>
      )

      return (
        <DeactivatablePopout direction='right'
                             className='task-status-popout'
                             control={popoutButton}>
          <div className="popout-content__header active-task-details--bordered">
            <h3 className="info-popout--name">
              <FormattedMessage {...messages.statusPrompt} />
            </h3>
          </div>

          <div className="popout-content__body">
            <span className="task-status__label">
              <FormattedMessage {...messagesByStatus[this.props.task.status]} />
            </span>
          </div>
        </DeactivatablePopout>
      )
    }
    else {
      return (
        <div className={classNames("task-status",
                                   "active-task-controls__vertical-control-block",
                                   "active-task-details--bordered",
                                   this.props.className)}>
          <div className="active-task-details--sub-heading">
            <FormattedMessage {...messages.statusPrompt} />
          </div>
          <div>
            <div className="task-status__label">
              <FormattedMessage {...messagesByStatus[this.props.task.status]} />
            </div>
          </div>
        </div>
      )
    }
  }
}

TaskStatusIndicator.propTypes = {
  task: PropTypes.object.isRequired,
  isMinimized: PropTypes.bool,
}
