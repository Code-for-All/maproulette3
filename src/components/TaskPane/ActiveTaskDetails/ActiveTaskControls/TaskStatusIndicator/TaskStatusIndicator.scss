@import '../../../../../variables.scss';

.active-task-details .task-status {
  margin-bottom: 15px;

  svg {
    height: 18px;
    width: auto;
    fill: $grey-dark;
    margin-right: 7px;
    vertical-align: text-top;
  }

  .task-status__prompt {
    font-weight: $weight-medium;
  }

  &.icon-only {
    .control-icon svg {
      fill: $yellow;
    }
  }
}
