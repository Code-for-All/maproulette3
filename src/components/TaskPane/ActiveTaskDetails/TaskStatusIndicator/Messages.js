import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with TaskStatusIndicator
 */
export default defineMessages({
  statusPrompt: {
    id: 'ActiveTask.status.prompt',
    defaultMessage: "Existing Status",
  },

  statusTooltip: {
    id: 'ActiveTask.status.control.toolTip',
    defaultMessage: "Existing Status",
  }
})
