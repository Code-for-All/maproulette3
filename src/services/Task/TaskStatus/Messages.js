import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with TaskStatus.
 */
export default defineMessages({
  created: {
    id: "Task.status.created",
    defaultMessage: "Created"
  },
  fixed: {
    id: "Task.status.fixed",
    defaultMessage: "Fixed"
  },
  falsePositive: {
    id: "Task.status.falsePositive",
    defaultMessage: "Not an Issue"
  },
  skipped: {
    id: "Task.status.skipped",
    defaultMessage: "Skipped"
  },
  deleted: {
    id: "Task.status.deleted",
    defaultMessage: "Deleted"
  },
  alreadyFixed: {
    id: "Task.status.alreadyFixed",
    defaultMessage: "Already Fixed"
  },
  tooHard: {
    id: "Task.status.tooHard",
    defaultMessage: "Too Hard"
  },
})
