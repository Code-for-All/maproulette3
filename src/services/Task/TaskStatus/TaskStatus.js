import { map as _map,
         invert as _invert,
         fromPairs as _fromPairs } from 'lodash'
import messages from './Messages'

// These statuses are defined on the server
export const TASK_STATUS_CREATED = 0
export const TASK_STATUS_FIXED = 1
export const TASK_STATUS_FALSE_POSITIVE = 2
export const TASK_STATUS_SKIPPED = 3
export const TASK_STATUS_DELETED = 4
export const TASK_STATUS_ALREADY_FIXED = 5
export const TASK_STATUS_TOO_HARD = 6

export const TaskStatus = Object.freeze({
  created: TASK_STATUS_CREATED,
  fixed: TASK_STATUS_FIXED,
  falsePositive: TASK_STATUS_FALSE_POSITIVE,
  skipped: TASK_STATUS_SKIPPED,
  deleted: TASK_STATUS_DELETED,
  alreadyFixed: TASK_STATUS_ALREADY_FIXED,
  tooHard: TASK_STATUS_TOO_HARD,
})

export const keysByStatus = Object.freeze(_invert(TaskStatus))

/**
 * Returns true if the given status represents completion of
 * the task.
 */
export const isCompleted = function(status) {
  return status === TaskStatus.fixed ||
         status === TaskStatus.falsePositive ||
         status === TaskStatus.alreadyFixed ||
         status === TaskStatus.deleted
}

/**
 * Returns an object mapping difficulty values to raw internationalized
 * messages suitable for use with FormattedMessage or formatMessage.
 */
export const messagesByStatus = _fromPairs(
  _map(messages, (message, key) => [TaskStatus[key], message])
)

/** Returns object containing localized labels  */
export const statusLabels = intl => _fromPairs(
  _map(messages, (message, key) => [key, intl.formatMessage(message)])
)
