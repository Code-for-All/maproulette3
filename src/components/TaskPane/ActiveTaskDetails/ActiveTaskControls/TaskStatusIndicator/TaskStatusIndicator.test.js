import React from 'react'
import TaskStatusIndicator from './TaskStatusIndicator'
import { TaskStatus } from '../../../../../services/Task/TaskStatus/TaskStatus'

let basicProps = null

beforeEach(() => {
  basicProps = {
    task: {
      id: 123,
      parent: {
        id: 321,
      },
      status: TaskStatus.created,
    },
    intl: {formatMessage: jest.fn(m => m.defaultMessage)},
    isMinimized: false,
  }
})

test("shows the task status", () => {
  const wrapper = shallow(
    <TaskStatusIndicator {...basicProps} />
  )

  expect(wrapper.find('.task-status').exists()).toBe(true)

  expect(wrapper).toMatchSnapshot()
})

test("shows popout control when minimized", () => {
  basicProps.isMinimized = true

  const wrapper = shallow(
    <TaskStatusIndicator {...basicProps} />
  )

  expect(wrapper.find('.task-status-popout').exists()).toBe(true)

  expect(wrapper).toMatchSnapshot()
})
