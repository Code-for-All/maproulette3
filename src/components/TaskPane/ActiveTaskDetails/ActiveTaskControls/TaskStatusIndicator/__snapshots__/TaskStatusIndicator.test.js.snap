// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shows popout control when minimized 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskStatusIndicator
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
    isMinimized={true}
    task={
        Object {
            "id": 123,
            "parent": Object {
              "id": 321,
            },
            "status": 0,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <div
          className="popout-content__header active-task-details--bordered"
>
          <h3
                    className="info-popout--name"
          >
                    <FormattedMessage
                              defaultMessage="Existing Status"
                              id="ActiveTask.status.prompt"
                              values={Object {}}
                    />
          </h3>
</div>,
        <div
          className="popout-content__body"
>
          <span
                    className="task-status__label"
          >
                    <FormattedMessage
                              defaultMessage="Created"
                              id="Task.status.created"
                              values={Object {}}
                    />
          </span>
</div>,
      ],
      "className": "task-status-popout",
      "control": <button
        className="button icon-only task-status"
>
        <span
                className="control-icon"
                title="Existing Status"
        >
                <SvgSymbol
                        sym="flag-icon"
                        viewBox="0 0 20 20"
                />
        </span>
</button>,
      "direction": "right",
      "eventTypes": Array [
        "mousedown",
        "touchstart",
      ],
      "excludeScrollbar": false,
      "outsideClickIgnoreClass": "ignore-react-onclickoutside",
      "preventDefault": false,
      "stopPropagation": false,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <h3
            className="info-popout--name"
>
            <FormattedMessage
                        defaultMessage="Existing Status"
                        id="ActiveTask.status.prompt"
                        values={Object {}}
            />
</h3>,
          "className": "popout-content__header active-task-details--bordered",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <FormattedMessage
              defaultMessage="Existing Status"
              id="ActiveTask.status.prompt"
              values={Object {}}
/>,
            "className": "info-popout--name",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "defaultMessage": "Existing Status",
              "id": "ActiveTask.status.prompt",
              "values": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "h3",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <span
            className="task-status__label"
>
            <FormattedMessage
                        defaultMessage="Created"
                        id="Task.status.created"
                        values={Object {}}
            />
</span>,
          "className": "popout-content__body",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <FormattedMessage
              defaultMessage="Created"
              id="Task.status.created"
              values={Object {}}
/>,
            "className": "task-status__label",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "defaultMessage": "Created",
              "id": "Task.status.created",
              "values": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "span",
        },
        "type": "div",
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <div
            className="popout-content__header active-task-details--bordered"
>
            <h3
                        className="info-popout--name"
            >
                        <FormattedMessage
                                    defaultMessage="Existing Status"
                                    id="ActiveTask.status.prompt"
                                    values={Object {}}
                        />
            </h3>
</div>,
          <div
            className="popout-content__body"
>
            <span
                        className="task-status__label"
            >
                        <FormattedMessage
                                    defaultMessage="Created"
                                    id="Task.status.created"
                                    values={Object {}}
                        />
            </span>
</div>,
        ],
        "className": "task-status-popout",
        "control": <button
          className="button icon-only task-status"
>
          <span
                    className="control-icon"
                    title="Existing Status"
          >
                    <SvgSymbol
                              sym="flag-icon"
                              viewBox="0 0 20 20"
                    />
          </span>
</button>,
        "direction": "right",
        "eventTypes": Array [
          "mousedown",
          "touchstart",
        ],
        "excludeScrollbar": false,
        "outsideClickIgnoreClass": "ignore-react-onclickoutside",
        "preventDefault": false,
        "stopPropagation": false,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h3
              className="info-popout--name"
>
              <FormattedMessage
                            defaultMessage="Existing Status"
                            id="ActiveTask.status.prompt"
                            values={Object {}}
              />
</h3>,
            "className": "popout-content__header active-task-details--bordered",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <FormattedMessage
                defaultMessage="Existing Status"
                id="ActiveTask.status.prompt"
                values={Object {}}
/>,
              "className": "info-popout--name",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "defaultMessage": "Existing Status",
                "id": "ActiveTask.status.prompt",
                "values": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "h3",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <span
              className="task-status__label"
>
              <FormattedMessage
                            defaultMessage="Created"
                            id="Task.status.created"
                            values={Object {}}
              />
</span>,
            "className": "popout-content__body",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <FormattedMessage
                defaultMessage="Created"
                id="Task.status.created"
                values={Object {}}
/>,
              "className": "task-status__label",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "defaultMessage": "Created",
                "id": "Task.status.created",
                "values": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "span",
          },
          "type": "div",
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`shows the task status 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TaskStatusIndicator
    intl={
        Object {
            "formatMessage": [Function],
          }
    }
    isMinimized={false}
    task={
        Object {
            "id": 123,
            "parent": Object {
              "id": 321,
            },
            "status": 0,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="active-task-details--sub-heading"
>
          <FormattedMessage
                    defaultMessage="Existing Status"
                    id="ActiveTask.status.prompt"
                    values={Object {}}
          />
</div>,
        <div>
          <div
                    className="task-status__label"
          >
                    <FormattedMessage
                              defaultMessage="Created"
                              id="Task.status.created"
                              values={Object {}}
                    />
          </div>
</div>,
      ],
      "className": "task-status active-task-controls__vertical-control-block active-task-details--bordered",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <FormattedMessage
            defaultMessage="Existing Status"
            id="ActiveTask.status.prompt"
            values={Object {}}
/>,
          "className": "active-task-details--sub-heading",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "defaultMessage": "Existing Status",
            "id": "ActiveTask.status.prompt",
            "values": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="task-status__label"
>
            <FormattedMessage
                        defaultMessage="Created"
                        id="Task.status.created"
                        values={Object {}}
            />
</div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <FormattedMessage
              defaultMessage="Created"
              id="Task.status.created"
              values={Object {}}
/>,
            "className": "task-status__label",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "defaultMessage": "Created",
              "id": "Task.status.created",
              "values": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="active-task-details--sub-heading"
>
            <FormattedMessage
                        defaultMessage="Existing Status"
                        id="ActiveTask.status.prompt"
                        values={Object {}}
            />
</div>,
          <div>
            <div
                        className="task-status__label"
            >
                        <FormattedMessage
                                    defaultMessage="Created"
                                    id="Task.status.created"
                                    values={Object {}}
                        />
            </div>
</div>,
        ],
        "className": "task-status active-task-controls__vertical-control-block active-task-details--bordered",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <FormattedMessage
              defaultMessage="Existing Status"
              id="ActiveTask.status.prompt"
              values={Object {}}
/>,
            "className": "active-task-details--sub-heading",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "defaultMessage": "Existing Status",
              "id": "ActiveTask.status.prompt",
              "values": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="task-status__label"
>
              <FormattedMessage
                            defaultMessage="Created"
                            id="Task.status.created"
                            values={Object {}}
              />
</div>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <FormattedMessage
                defaultMessage="Created"
                id="Task.status.created"
                values={Object {}}
/>,
              "className": "task-status__label",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "defaultMessage": "Created",
                "id": "Task.status.created",
                "values": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
